core/embed/rust/src/micropython/gc.rs

```diff
@@ -0,0 +1,79 @@
+use core::{
+    alloc::Layout,
+    marker::Unsize,
+    ops::{CoerceUnsized, Deref, DispatchFromDyn},
+    ptr::{self, NonNull},
+};
+
+use super::ffi;
+
+/// A pointer type for values on the garbage-collected heap.
+///
+/// Although a garbage-collected pointer type technically should implement
+/// `Copy` and `Clone`, we avoid doing this until proven necessary.
+pub struct Gc<T: ?Sized>(NonNull<T>);
+
+impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Gc<U>> for Gc<T> {}
+impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Gc<U>> for Gc<T> {}
+
+impl<T> Gc<T> {
+    /// Allocate memory on the heap managed by the MicroPython garbage collector
+    /// and then place `v` into it. `v` will _not_ get its destructor called.
+    pub fn new(v: T) -> Self {
+        let layout = Layout::for_value(&v);
+        // TODO: Assert that `layout.align()` is the same as the GC alignment.
+        // SAFETY:
+        //  - Unfortunately we cannot respect `layout.align()` as MicroPython GC does
+        //    not support custom alignment.
+        //  - `ptr` is guaranteed to stay valid as long as it's reachable from the stack
+        //    or the MicroPython heap.
+        unsafe {
+            let raw = ffi::gc_alloc(layout.size(), 0).cast();
+            ptr::write(raw, v);
+            Self::from_raw(raw)
+        }
+    }
+
+    /// Return a mutable reference to the value.
+    ///
+    /// # Safety
+    ///
+    /// `Gc` values can originate in the MicroPython interpreter, and these can
+    /// be both shared and mutable. Before calling this function, you have to
+    /// ensure that `this` is unique for the whole lifetime of the
+    /// returned mutable reference.
+    pub unsafe fn as_mut(this: &mut Self) -> &mut T {
+        // SAFETY: The caller must guarantee that `this` meets all the requirements for
+        // a mutable reference.
+        unsafe { this.0.as_mut() }
+    }
+}
+
+impl<T: ?Sized> Gc<T> {
+    /// Construct a `Gc` from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because the caller has to guarantee that `ptr`
+    /// is pointing to a memory previously allocated on the GC heap, i.e.
+    /// through `Gc::new()`, or through the MicroPython interpreter.
+    pub unsafe fn from_raw(ptr: *mut T) -> Self {
+        // SAFETY: The caller must guarantee that `ptr` points to a previously
+        // GC-allocated memory.
```

Comment:
...and explain _here_ that `Gc` is also allowed to wrap ... well ... anything the Micropython GC can reason about, basically? so context pointers too?